pkg/core/statesync/module.go

```diff
@@ -0,0 +1,462 @@
+	// none means that state exchange is enabled in the configuration, but
+	// initialisation of the state sync module wasn't yet performed, i.e.
+	// (*Module).Init wasn't called. Can't be combined with other states.
+	none
+	// initialized means that (*Module).Init was called, but other sync stages
+	// are not yet reached (i.e. that headers are requested, but not yet fetched).
+	// Can't be combined with other states.
+	initialized
+	// headersSynced means that headers for the current state sync point are fetched.
+	// May be combined with mptSynced and/or blocksSynced.
+	headersSynced
+	// mptSynced means that MPT nodes for the current state sync point are fetched.
+	// Always combined with headersSynced; may be combined with blocksSynced.
+	mptSynced
+	// blocksSynced means that blocks up to the current state sync point are stored.
+	// Always combined with headersSynced; may be combined with mptSynced.
+	blocksSynced
+)
+
+// Module represents state sync module and aimed to gather state-related data to
+// perform an atomic state jump.
+type Module struct {
+	lock sync.RWMutex
+	log  *zap.Logger
+
+	// syncPoint is the state synchronisation point P we're currently working against.
+	syncPoint uint32
+	// syncStage is the stage of the sync process.
+	syncStage stateSyncStage
+	// syncInterval is the delta between two adjacent state sync points.
+	syncInterval uint32
+	// blockHeight is the index of the latest stored block.
+	blockHeight uint32
+
+	dao     *dao.Simple
+	bc      blockchainer.Blockchainer
+	mptpool *Pool
+
+	billet          *mpt.Billet
+	requestMPTNodes func(nodes map[util.Uint256]struct{}) error
+}
+
+// NewModule returns new instance of statesync module.
+func NewModule(bc blockchainer.Blockchainer, log *zap.Logger, s *dao.Simple) *Module {
+	if !(bc.GetConfig().P2PStateExchangeExtensions && bc.GetConfig().RemoveUntraceableBlocks) {
+		return &Module{
+			dao:       s,
+			bc:        bc,
+			syncStage: inactive,
+		}
+	}
+	return &Module{
+		dao:          s,
+		bc:           bc,
+		log:          log,
+		syncInterval: uint32(bc.GetConfig().StateSyncInterval),
+		mptpool:      NewPool(),
+		syncStage:    none,
+		billet:       mpt.NewBillet(bc.GetConfig().KeepOnlyLatestState, s.Store),
+	}
+}
+
+// Init initializes state sync module for the current chain's height with given
+// callback for MPT nodes requests.
+func (s *Module) Init(currChainHeight uint32, reqMPTNodes func(nodes map[util.Uint256]struct{}) error) error {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	if s.syncStage != none {
+		return nil
+	}
+	s.requestMPTNodes = reqMPTNodes
+
+	p := (currChainHeight / s.syncInterval) * s.syncInterval
+	if p < 2*s.syncInterval {
+		// chain is too low to start state exchange process, use the standard sync mechanism
+		s.syncStage = inactive
+		return nil
+	}
+	pOld, err := s.dao.GetStateSyncPoint()
+	if err == nil && pOld >= p-s.syncInterval {
+		// old point is still valid, so try to resync states for this point.
```

Comment:
Fixed for blocks\transactions\AERs. For NEP17 transfers `TODO` comment with a link to https://github.com/nspcc-dev/neo-go/issues/2096 created.

And one more time I've faced with MPT versioning problem: we can't modify existing (even outdated) MPT state, it's the one current Blockchain state is based on. We can't just remove outdated MPT nodes, because no one guarantees that currently running state sync process will be successfully completed. Instead, we should remove outdated MPT state along with outdated blocks\transactions\AERs right inside (Blockchain).JumpToState. So I like the idea of MPT versioning more and more.